client/dashboard/app/i18n.tsx

```diff
@@ -0,0 +1,55 @@
+import { defaultI18n } from '@wordpress/i18n';
+import { I18nProvider as WPI18nProvider } from '@wordpress/react-i18n';
+import { useEffect, useRef, useState, type PropsWithChildren } from 'react';
+import { useAuth } from './auth';
+
+async function fetchLocaleData( language: string ) {
+	if ( language === 'en' ) {
+		return [ language, undefined ];
+	}
+
+	try {
+		const response = await fetch(
+			`https://widgets.wp.com/languages/calypso/${ language }-v1.1.json`
+		);
+
+		return [ language, await response.json() ];
+	} catch {
+		// Fall back to `en` when fetching the language data fails. Without this
+		// the i18n provider would be stuck forever in a non-loaded state.
+		return [ 'en', undefined ];
+	}
+}
+
+export function I18nProvider( { children }: PropsWithChildren ) {
+	const [ loadedLocale, setLoadedLocale ] = useState< string | null >( null );
+	const loadingLocaleRef = useRef< string | null >( null );
+	const { user } = useAuth();
+
+	const i18n = defaultI18n;
+	const language = user.locale_variant || user.language || 'en';
+
+	useEffect( () => {
+		loadingLocaleRef.current = language;
+
+		fetchLocaleData( language ).then( ( [ realLanguage, data ] ) => {
```

Comment:
It's unlikely, but should this component be resilient to being unmounted, and handling what happens if it's unmounted while the fetch is still in-flight? (e.g. something with `AbortController` or maybe unsetting `loadingLocaleRef.current` in the `useEffect` return value function).

Reply:
Good point, I rewrote the effect to use `AbortController`. Now it should handle all the relevant situations much better: abort when another locale switch is initiated while the previous one hasn't finished fetching yet; abort also when umounting and don't call `i18n` methods after the component was unmounted.